### PR TITLE
[X] Correlations: Allow creating correlations for provisioned data sources

### LINKED ISSUE #71110: Glue: Allow adding correlations for read-only data sources
__Background__

Design doc: https://docs.google.com/document/d/1BjJ1GDxY3BjFzF9ZZKe19gN7foyLDZa05k2U1YSlPyM/edit
Proposal: 1

__Requirements__

1. Remove validation code that prevents adding a correlation to a read-only data source (e.g. [source](https://github.com/grafana/grafana/blob/v10.0.1/pkg/services/correlations/database.go#L35) but applicable to update and delete as well)
2. Keep automatic removals when a data source is removed (this requires extra logic when a provisioned data source is re-created, see Technical Details below)
3. Provisioned correlations should be read-only. This requires changing how read-only is read in the frontend, it will require returning the `provisioned` property in the API (e.g. [source](https://github.com/grafana/grafana/blob/v10.0.1/public/app/features/correlations/CorrelationsPage.tsx#L33))
4. Update documentation (API and User docs)

Post-merge testing:
- Test in Cloud that renaming a stack keeps correlations intact 

__Technical details__

- Add `provisioned` column to database (alternatively, create a separate table like Dashboards and Alerts)
- Manually track what data sources are re-created during provisioning to clean-up unused resources

<details><summary>Provisioning algorithm</summary>

Check the design doc for more details. 

https://github.com/grafana/grafana/blob/v10.0.1/pkg/services/provisioning/datasources/datasources.go#L60 

1. Skip firing an event when the data source is deleted. Ideally: return a list of events when `dc.deleteDatasources` is called in provisioning, filter out data sources that got re-created, and fire events again to perform the cleanup.
2. Before provisioning data sources: remove all correlations where column `provisioned=true`
3. Create provisioned correlations as normal

</details>

_Suggested task split_

- [ ] Requirements: 2, 3. Add the `provisioning` column and change the provisioning logic. This can be tested by changing `provisioned` column in the db
- [ ] Requirements: 1, 4. Remove API validation, update frontend code, update documentation

### PR DESCRIPTION
**Pull request by ifrost**
Fixes #71110

This is to allow creating correlations for provisioned data sources. Please check  #71110 for more details and design doc.

Key changes:
* Clean-up logic was adjusted to ensure correlations are removed when a data source is removed, taking into account a scenario when a data source is re-created during provisioning and no data loss occurs
* Previously, correlations attached to a read-only data source were marked as read-only. Now, only provisioned correlations are read-only 

**Special notes for your reviewer:**

<details><summary>How to test it?</summary>

You can use your own/current setup or follow manual test scenario I've created: https://gist.github.com/ifrost/35b7d7424a625d9cd8fc8b845fbd002a

Try out different things:
* Adding new correlations to different data sources (provisioned and not provisioned)
* Removing manually added correlations
* Add `deleteDatasources` section in provisioning config with different datasources that are being added later; then try actually removing some of the data sources
* Adding/Editing/Removing correlations inside provisioning file
* Restart Grafana with the same config

After each step, you should see correlations belonging to existing data sources. Manual entries should not be removed during provisioning. Check the sqlite db directly, not only frontend (we filter out correlations for data sources that do not exist)

</details>

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/correlations/models.go [hunk 1]
```diff
@@ -9,7 +9,7 @@ import (
 )
 
 var (
-	ErrSourceDataSourceReadOnly           = errors.New("source data source is read only")
+	ErrCorrelationsReadOnly               = errors.New("correlation is read only")
```
Comment (Elfo404) [inline pkg/services/correlations/models.go:12]: maybe we should use the singular form?
```suggestion
	ErrCorrelationReadOnly               = errors.New("correlation is read only")
```

File: pkg/services/correlations/api.go [hunk 2]
```diff
@@ -56,8 +56,8 @@ func (s *CorrelationsService) createHandler(c *contextmodel.ReqContext) response
 			return response.Error(http.StatusNotFound, "Data source not found", err)
 		}
 
-		if errors.Is(err, ErrSourceDataSourceReadOnly) {
-			return response.Error(http.StatusForbidden, "Data source is read only", err)
+		if errors.Is(err, ErrCorrelationsReadOnly) {
+			return response.Error(http.StatusForbidden, "Correlations is read only", err)
```
Comment (Elfo404) [inline pkg/services/correlations/api.go:60]: how can this happen? i'm creating a new one, how can i get a read-only error?

File: pkg/services/provisioning/datasources/datasources.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ func newDatasourceProvisioner(log log.Logger, store Store, correlationsStore Cor
 }
 
 func (dc *DatasourceProvisioner) apply(ctx context.Context, cfg *configs) error {
-	if err := dc.deleteDatasources(ctx, cfg.DeleteDatasources); err != nil {
+	if err := dc.deleteDatasources(ctx, cfg.DeleteDatasources, cfg.Datasources); err != nil {
```
Comment (Elfo404) [inline pkg/services/provisioning/datasources/datasources.go:61]: this is pretty hard to follow imho, can we move the diff logic outside of `deleteDatasources` so that it only deletes without performing any extra logic that does not belong there?

we could probably create an intermediary slice that also contains `SkipPublish` and simply use that in`deleteDatasources`

Comment (ifrost) [reply pkg/services/provisioning/datasources/datasources.go:61]: > so that it only deletes without performing any extra logic that does not belong there?

We still need to perform clean-up logic for data sources that were actually deleted. We cannot do it by removing correlations directly in the provisioning code, we need to actually publish DeleteDatasource event which triggers the clean-up at module level. This is needed not only to avoid code duplication but to support cases when more clean-up handlers are added in the code.

My first thought was that we can pass SkipPublish flag to the wrapper that deletes all data sources. When SkipPublish is set it would return a list of events that has to be published manually byt the consumer at the end. During provisioning we can remove from that list events for data sources that were not deleted, but re-created. I just thought it would actually harder to follow (with skip flag pushed farther down to the store), but I can try it out. What do you think?

File: pkg/services/correlations/models.go (hunk 1 above)
Comment (ifrost) [reply pkg/services/correlations/models.go]: Good call! Updated in 233e13fb10e5c76d1234a2430377e2bddff11095

File: pkg/services/correlations/api.go (hunk 2 above)
Comment (ifrost) [reply pkg/services/correlations/api.go]: Oh, true that! Removed the check in 1806646de5837faebac97526eff1cf82dac90daa. Is there a neat way to lint or annotate such cases? Something that would work like good old `try { } catch(ErrorA a) {} catch(ErrorB b) {}`

File: pkg/services/provisioning/datasources/datasources.go (hunk 3 above)
Comment (Elfo404) [reply pkg/services/provisioning/datasources/datasources.go:61]: i see what you mean and it makes sense, i tried moving this logic out but then you actually need to query the DB to get the datasource UID, and we probably only want to do that in the delete function otherwise a lot more refactoring would be needed.

however i can se a case in which this doesn't work: when the datasources to be added do not contain an explicit UID, when doing `uidsToBeAdded = append(uidsToBeAdded, ds.UID)` we append a bunch of `""`, but the UID generation is actually stable and the datasource is effectively being recreated, but we treat it as only deleted

Comment (ifrost) [reply pkg/services/provisioning/datasources/datasources.go:61]: Good catch!

I can use `safeUIDFromName` to figure out missing UIDs  like:

```go
var uidsToBeAdded []string
for _, ds := range dsToBeAdded {
    if ds.UID != "" {
	uidsToBeAdded = append(uidsToBeAdded, ds.UID)
    } else {
	uidsToBeAdded = append(uidsToBeAdded, safeUIDFromName(ds.Name))
    }
}
```

An alternative option is to change the provisioning logic so it first generates a list off commands to be executed. Then when it invokes `DeleteDataSourceCommand` it passes SkipPublish if there's an insert coming down the list. That would require changing deletion a bit so it generates UID like `createInsertCommand` but I'd need to check what are other implications of doing that.

**Comment by ifrost** [comment]
New PR: https://github.com/grafana/grafana/pull/71984
